### ISSUE TITLE
Possible bug on 3.1.2-SNAPSHOT

### ISSUE THREAD

**Issue by felislynx-silae** [Samsung/GearVRf#1104]
Ever since i've made update to 3.1.2 (for daydream support), i've random exceptions like this on gear
```
Exception java.util.ConcurrentModificationException:
java.util.ArrayList$ArrayListIterator.next (ArrayList.java:573)
org.gearvrf.GVRBaseSensor.processList (GVRBaseSensor.java:134)
org.gearvrf.SensorManager.processPick (SensorManager.java:88)
org.gearvrf.GVRCursorController.update (GVRCursorController.java:618)
org.gearvrf.GVRCursorController.setPosition (GVRCursorController.java:423)
org.gearvrf.io.GVRGazeCursorController.access$101 (GVRGazeCursorController.java:30)
org.gearvrf.io.GVRGazeCursorController$EventHandlerThread$1.handleMessage (GVRGazeCursorController.java:249)
android.os.Handler.dispatchMessage (Handler.java:98)
android.os.Looper.loop (Looper.java:158)
android.os.HandlerThread.run (HandlerThread.java:61)
```

**Comment by liaxim** [comment]
@felislynx-silae Do you have any hints on how to reproduce? Does it happen with any of our demo apps? Thanks.

@rahul27 Let me know if you are too busy. I could look into it.

**Comment by senseearth** [comment]
@felislynx-silae Hi there. Would be happy to help you with it. To help us reproduce, it would be great to share with us

1. A simplified demo project with source code
2. Steps to have this occur, and frequency, and if you are doing some multithreading as well
3. And other hardware factors such as is it gearvr only or daydream as well, and the particular device you use

Thanks

**Comment by felislynx-silae** [comment]
1. I'll have to prepare demo project with sourcecode i could share
2. I'ts on startup of app (sometimes on going back to this scene) where i have 12 sceneobjects with selectable behavior (one on the middle is receiving event.isOver() event even if i'm not looking at it) only multithreading that is happening is retrofit call to our server which is loading data.
3. I didn't checked on daydream since this version is on early stage of development. Reference device is Galaxy S6, Android 6.0.1 most up to date oculus drivers. It happens on second S6 and on S7 with 7.0 Android

**Comment by rahul27** [comment]
@liaxim I can take a look at it when @felislynx-silae shares the demo code.

**Comment by rahul27** [comment]
Hi @felislynx-silae, do you have an update for us?

Also is this something you see with a stock sample as well?

**Comment by felislynx-silae** [comment]
@rahul27 sorry for the delay. I need to scratch some things from my timetable in work before i can go back to this. I beleive in next week i'll publish sample project on github for you. Again, sorry for delay but you know, deadlines.

**Comment by rahul27** [comment]
@felislynx-silae Sure. No issues :smile:

**Comment by rahul27** [comment]
@felislynx-silae I tried recreating the issue based on the information you provided but was unable to reproduce the exception. I created a sample with multiple objects/selectable behaviors to test.

I have a few questions that might help me understand the issue better:

1. Does this happen every time you start the app or is it intermittent?
2. Are you only using the gaze cursor in your application? 
3. You mentioned an issue with event.isOver(), do you also have a CursorEventListener registered in your application?

**Comment by felislynx-silae** [comment]
1. It is intermittent. It happens like few times a day
2. No, it's laser cursor
3. Yes, i'm listening to events that are happening to those scene objects. This screen is a menu so tiles reacts to hover and click events.

I've started on sample app which will recreate bug.

**Comment by felislynx-silae** [comment]
Headsup: https://github.com/felislynx-silae/GearBugSample
```
03-31 08:06:29.866 8318-8353/eu.lynxit.sampleapp E/AndroidRuntime: FATAL EXCEPTION: GVRGazeEventHandlerThread
                                                                   Process: eu.lynxit.sampleapp, PID: 8318
                                                                   java.util.ConcurrentModificationException
                                                                       at java.util.ArrayList$Itr.next(ArrayList.java:831)
                                                                       at org.gearvrf.GVRBaseSensor.processList(GVRBaseSensor.java:134)
                                                                       at org.gearvrf.SensorManager.processPick(SensorManager.java:88)
                                                                       at org.gearvrf.GVRCursorController.update(GVRCursorController.java:618)
                                                                       at org.gearvrf.GVRCursorController.setPosition(GVRCursorController.java:423)
                                                                       at org.gearvrf.io.GVRGazeCursorController.access$101(GVRGazeCursorController.java:30)
                                                                       at org.gearvrf.io.GVRGazeCursorController$EventHandlerThread$1.handleMessage(GVRGazeCursorController.java:249)
                                                                       at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                       at android.os.Looper.loop(Looper.java:154)
                                                                       at android.os.HandlerThread.run(HandlerThread.java:61)
```

Steps 2 Reproduce :
1. Open app and look into thumbnail
2. If it doesn't crash almost immediately, go to step 1. (it might take a while)

**Comment by liaxim** [comment]
@rahul27 Very likely the problem is originating from the fact this app is installing a new main scene. GVRInputManagerImpl.setScene would get called on the framework thread unlike the usual GVRGazeEventHandlerThread.

**Comment by felislynx-silae** [comment]
By the way. Today i've discovered new type of error (it wasn't present before today). I've checked on other phones where i had build made before today and bug was not present (no code changes made to project)
```
04-04 10:12:39.971 3275-3416/eu.lynxit.sampleapp E/AndroidRuntime: FATAL EXCEPTION: GVRGazeEventHandlerThread
                                                                   Process: eu.lynxit.sampleapp, PID: 3275
                                                                   java.lang.NullPointerException: Attempt to invoke virtual method 'org.gearvrf.GVRComponent org.gearvrf.GVRSceneObject.getComponent(long)' on a null object reference
                                                                       at org.gearvrf.io.cursor3d.CursorManager.callEventHandler(CursorManager.java:1111)
                                                                       at org.gearvrf.io.cursor3d.CursorManager.access$1300(CursorManager.java:67)
                                                                       at org.gearvrf.io.cursor3d.CursorManager$4.onEvent(CursorManager.java:1105)
                                                                       at org.gearvrf.io.cursor3d.Cursor.dispatchCursorEvent(Cursor.java:616)
                                                                       at org.gearvrf.io.cursor3d.Cursor.handleControllerEvent(Cursor.java:666)
                                                                       at org.gearvrf.io.cursor3d.LaserCursor$1.onEvent(LaserCursor.java:117)
                                                                       at org.gearvrf.GVRCursorController.update(GVRCursorController.java:620)
                                                                       at org.gearvrf.GVRCursorController.setPosition(GVRCursorController.java:423)
                                                                       at org.gearvrf.io.GVRGazeCursorController.access$101(GVRGazeCursorController.java:30)
                                                                       at org.gearvrf.io.GVRGazeCursorController$EventHandlerThread$1.handleMessage(GVRGazeCursorController.java:254)
                                                                       at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                       at android.os.Looper.loop(Looper.java:158)
                                                                       at android.os.HandlerThread.run(HandlerThread.java:61)
```
S2R:
- Build project
- click (volume buttons, touch screen or on headset) somewhere on scene (for example below all thumbnails)

**Comment by rahul27** [comment]
@felislynx-silae Yes. I saw the issue while trying your app. I should have a fix out for you soon.

**Comment by rahul27** [comment]
@felislynx-silae I've added a fix to the framework : 

https://github.com/Samsung/GearVRf/pull/1128

I've also updated your sample to include the newly compiled framework libraries with the fix. You can find it here:

https://github.com/felislynx-silae/GearBugSample/pull/1

Could you try it an let me know if it fixes your issue.

**Comment by felislynx-silae** [comment]
As far as i can tell, it's fixed both crashes. I've tried to reproduce this bug for over hour on 2 phones. If anything happens on them until end of day, i'll let you know

**Comment by liaxim** [comment]
Fixed in 3.2
